### PR TITLE
Fixed #24375 -- Added Migration.initial property

### PR DESCRIPTION
**Pull request by akulakov**
Added Migration.initial property to be check when
--fake-initial arg is used, initial will be set to
True for all initial migrations (this is particularly
useful when initial migrations are split). Users may
also set `initial = True` manually.

### CODE REVIEW

File: django/db/migrations/autodetector.py [hunk 1]
```diff
@@ -33,6 +33,7 @@ def __init__(self, from_state, to_state, questioner=None):
         self.from_state = from_state
         self.to_state = to_state
         self.questioner = questioner or MigrationQuestioner()
+        self.apps_not_initial = [app for app, model in from_state.models]
```
Comment (carljm) [inline django/db/migrations/autodetector.py]: May as well make this a set here, since it's used only for containment checks? No purpose in keeping dupe app names in it.

Also, I think the name `existing_apps` would be clearer (I didn't understand this name until I looked how it was used), but that's bikeshedding, so do with it what you will.

File: django/db/migrations/autodetector.py [hunk 2]
```diff
@@ -1051,6 +1053,7 @@ def arrange_for_graph(self, changes, graph, migration_name=None):
                     migration.dependencies.append(app_leaf)
                 if i == 0 and not app_leaf:
                     new_name = "0001_%s" % migration_name if migration_name else "0001_initial"
+                    migration.initial = True
```
Comment (carljm) [inline django/db/migrations/autodetector.py]: Hmm, I may not be fully understanding all this code, but am I understanding right that this is technically redundant, because the app should not be in the previous state at all, so it should have already had `initial` set to `True` in `_detect_changes()`? Or are there cases where that would not be the case?

File: django/db/migrations/executor.py [hunk 3]
```diff
@@ -117,10 +117,13 @@ def apply_migration(self, state, migration, fake=False, fake_initial=False):
         """
         Runs a migration forwards.
         """
+        # print ("in apply_migration()")
```
Comment (carljm) [inline django/db/migrations/executor.py]: This should be removed.

File: django/db/migrations/executor.py [hunk 4]
```diff
@@ -117,10 +117,13 @@ def apply_migration(self, state, migration, fake=False, fake_initial=False):
...
         if self.progress_callback:
             self.progress_callback("apply_start", migration, fake)
         if not fake:
+            # print('not fake')
```
Comment (carljm) [inline django/db/migrations/executor.py]: Also this.

File: django/db/migrations/executor.py [hunk 5]
```diff
@@ -117,10 +117,13 @@ def apply_migration(self, state, migration, fake=False, fake_initial=False):
...
             if fake_initial:
+                # print('fake initial')
```
Comment (carljm) [inline django/db/migrations/executor.py]: And this.

File: docs/topics/migrations.txt [hunk 6]
```diff
@@ -276,6 +276,14 @@ class to make it importable::
 Please refer to the notes about :ref:`historical-models` in migrations to see
 the implications that come along.
 
+.. _initial-property:
+
+.. note::
```
Comment (carljm) [inline docs/topics/migrations.txt]: The current structure makes it look like this is a note under the `Model Managers` heading, which doesn't make sense. I think it should just be a regular paragraph under a new `Initial migrations` header, instead of a callout note.

File: docs/topics/migrations.txt [hunk 7]
```diff
@@ -276,6 +276,14 @@ class to make it importable::
...
+
+    You may add an ``initial = True`` class property to the ``Migration`` class
+    to have Django apply the "initial migration" logic when the
```
Comment (carljm) [inline docs/topics/migrations.txt]: I think perhaps we may want a clearer description here than "apply the initial migration logic." Perhaps something like `... to have Django fake this migration if it finds its added tables or columns already present in the database when ...` ?

File: docs/topics/migrations.txt [hunk 8]
```diff
@@ -276,6 +276,14 @@ class to make it importable::
...
+    :djadminopt:`--fake-initial` argument is used.
```
Comment (carljm) [inline docs/topics/migrations.txt]: I'm tempted to append a sentence like `This is advanced functionality; if you aren't sure whether you need it, you probably don't.` to the end of this paragraph.

Comment (carljm) [inline docs/topics/migrations.txt]: We may also want to say explicitly here "If no `initial` class attribute is specified, Django will consider a migration to be an initial migration for an app if the migration has no dependencies in the same app."

File: docs/topics/migrations.txt (hunk 6 above)
Comment (carljm) [inline docs/topics/migrations.txt]: And we'll need a `versionchanged` marker on this section.

File: django/db/migrations/executor.py [hunk 9]
```diff
@@ -186,6 +194,16 @@ def detect_soft_applied(self, project_state, migration):
                 if model._meta.db_table not in self.connection.introspection.table_names(self.connection.cursor()):
                     return False, project_state
                 found_create_migration = True
-        # If we get this far and we found at least one CreateModel migration,
+
+            if isinstance(operation, migrations.AddField):
+                table = "%s_%s" % (migration.app_label, operation.model_name)
```
Comment (MarkusH) [inline django/db/migrations/executor.py]: You need to get the table out of the model this field belongs to. See the `isinstance(operation, migrations.CreateModel)` branch above.

File: django/db/migrations/executor.py [hunk 10]
```diff
@@ -186,6 +194,16 @@ def detect_soft_applied(self, project_state, migration):
...
+                fields = self.connection.introspection.get_table_description(self.connection.cursor(), table)
+                db_field = operation.name
+                if isinstance(operation.field, models.ForeignKey):
+                    db_field += "_id"
```
Comment (MarkusH) [inline django/db/migrations/executor.py]: You also need to get the name of the table column from the respective field

File: docs/topics/migrations.txt [hunk 11]
```diff
@@ -758,13 +766,11 @@ If you already have pre-existing migrations created with
 * Run ``python manage.py migrate --fake-initial``. Django will see that the
   tables for the initial migrations already exist and mark them as applied
   without running them. (Django won't check that the table schema match your
-  models, just that the right table names exist).
+  models, just that the right table names and field names exist).
```
Comment (carljm) [inline docs/topics/migrations.txt]: I think this overstates the capabilities of the detection, because it implies that field names are always checked, which is not true. For a `CreateModel` operation, even in this PR we still don't check that the field names match up if the table exists. We only check field names in the edge case of a separate `AddField` operation in a split initial migration.

I think it's important that we _not_ overstate this, so that people know they need to ensure the schema matches themselves. Rather than going into details, I would just leave the previous wording in place here.

**Comment by carljm** [comment]
Thanks for the PR; this is good work. I made some comments inline. Otherwise, the only thing I see is that the testing here doesn't look adequate yet. At the very least I think we need a new test demonstrating that a migration explicitly marked `initial = True` (even if not the first migration in the app) is faked with `--fake-initial`.

File: django/db/migrations/writer.py [hunk 12]
```diff
@@ -184,6 +185,9 @@ def as_string(self):
         if self.migration.replaces:
             items['replaces_str'] = "\n    replaces = %s\n" % self.serialize(self.migration.replaces)[0]
 
+        if self.migration.initial:
```
Comment (MarkusH) [inline django/db/migrations/writer.py]: `if self.migration.initial is not None` -- after all you set `False` explicitly.

File: docs/topics/migrations.txt (hunk 8 above)
Comment (MarkusH) [inline docs/topics/migrations.txt]: :+1: for both.

File: django/db/migrations/autodetector.py (hunk 2 above)
Comment (MarkusH) [inline django/db/migrations/autodetector.py]: I think this sets the `initial=True` for `0001_initial` while the other line above takes care of `0002_auto`.

**Comment by MarkusH** [comment]
Given the complexity of this change I'm worrying about the number of tests that it involves. There should:tm: be tests that explicitly check the faking of initial migrations; could you extend them and check for `initial=True` not just in the second but also in the first migration. If there aren't, now is the time when we should add them.

File: django/db/migrations/autodetector.py (hunk 2 above)
Comment (akulakov) [inline django/db/migrations/autodetector.py]: Carl's right -- when I started working on it, I started with the first migration which was easier to find in arrange_for_graphs(), then I realized I need to handle split migrations, which led me to add the code to _detect_changes(), which does cover all initial migrations; I later forgot to remove the code quoted above.

File: docs/topics/migrations.txt (hunk 7 above)
Comment (akulakov) [inline docs/topics/migrations.txt]: I think it gets hard to understand if it's kept as one sentence, might be better to have something like:

Django will fake the migration if these conditions are met: --fake-migrations arg is used, `initial = True` class property is added to migration class, and either of the two conditions are met: 1. migration creates a new table that already exists 2. migration adds new fields that already exist.

It may be more readable as a bullet list, especially because of the nested 'OR' sublist.

The tradeoff I see here is that this is a corner case that majority of readers don't need so does it need to take up so much real estate on this page?

Comment (akulakov) [inline docs/topics/migrations.txt]: Actually, if we want to explain the full logic in the easiest possible manner, how about something like this:

Django will fake a migration if the `--fake-migrations` arg is used, migration tries to create a new table or a new field that already exist, and either of the following conditions are met:

* migration has no dependencies

* migration has `initial = True` class property added to the migration class

[ note that it may also make sense to document the `initial = False` explicit skip of --fake-initial, if you can think of a case when this would be useful ]

File: docs/topics/migrations.txt (hunk 8 above)
Comment (akulakov) [inline docs/topics/migrations.txt]: I've addressed the 2nd addition in the comment above, I agree with the 1st addition, will do that..

File: django/db/migrations/writer.py (hunk 12 above)
Comment (akulakov) [inline django/db/migrations/writer.py]: I'm setting
                         instance.initial = app_label not in self.existing_apps
in autodetector, so when this is True, we need to write out `initial = True`, the tri-value is only in play when
creating migration from a migration file.

File: docs/topics/migrations.txt (hunk 7 above)
Comment (carljm) [inline docs/topics/migrations.txt]: I don't think this section will be very long regardless. This is the canonical location for explaining this behavior, so I think it's more important that we explain it clearly than that we keep it short. Here's my proposal:

> The "initial migrations" for an app are the migrations that create the first version of that app's tables. Usually an app will have just one initial migration, but in some cases of complex model interdependencies it may have two or more.
>
> Initial migrations are marked with an `initial = True` class attribute on the migration class. If no `initial` class attribute is found, a migration will be considered "initial" if it is the first migration in the app (i.e. it has no dependencies on any other migration in the same app).
>
> When :djadmin:`migrate` is run with the `--fake-initial` argument, these initial migrations are treated specially. For an initial migration that creates a one or more tables (`CreateModel` operation), Django will check if those tables already exist in the database, and fake-apply the migration if so. Similarly, for an initial migration that adds one or more fields (`AddField` operation), Django will check if those columns already exist in the database, and fake-apply the migration if so. When :djadminopt:`--fake-initial` is not used, initial migrations are treated no differently from any other migration.
>
> When auto-generating migrations, Django should always set the `initial` flag correctly. Modifying the flag manually is for advanced use cases only; if you aren't sure whether you need it, you probably don't.

File: django/db/migrations/writer.py (hunk 12 above)
Comment (carljm) [inline django/db/migrations/writer.py]: I think this code is correct as is; I'm not sure what @markush is suggesting. Initial can be tri valued, but regardless we only ever want to write `initial = True` into auto-written migration files.

**Comment by akulakov** [comment]
I have a question about testing.. 

First, to be clear, here's what's being tested now: 
1. when split initial migration instances are created, the second instance will be `initial=True`. (in my local changes I also test that 1st instance is `initial=True`)
2. when first initial migration file is written out, it will have the `initial = True` line.

Now we need a test that checks that a migration that HAS dependencies but also has initial = True, is faked on --fake-initial.

The easy way to do this would be to copy test_migrations dir to test_migrations_fake_initial, add initial = True to 2nd migration, modify test_commands.test_migrate_fake_initial() to use those tests & to test that running 0002 migration 2nd time with --fake-initial does not raise DatabaseError. 

Does that make sense?

**Comment by akulakov** [comment]
Another quick question, I've fixed some issues locally but not sure if I should push them because that would 'hide' some ongoing comments about documentation?

**Comment by carljm** [comment]
Pushing a new commit will only hide comments if it changes that exact line. And the comments are still available in the thread history. So I wouldn't worry about it.

Your suggestion for testing sounds okay to me.

**Comment by akulakov** [comment]
I've added the tests as discussed; By the way, when I push changes, should I add a comment here?

File: django/db/migrations/writer.py (hunk 12 above)
Comment (MarkusH) [inline django/db/migrations/writer.py]: Yes, I had to reread the patch again. The current way is correct.

File: docs/topics/migrations.txt [hunk 13]
```diff
@@ -276,6 +276,36 @@ class to make it importable::
...
+migration, but in some cases of complex model interdependencies it may have two
+or more.
+
+Initial migrations are marked with an ``initial = True`` class attribute on the
+migration class. If no initial class attribute is found, a migration will be
+considered "initial" if it is the first migration in the app (i.e. it has no
+dependencies on any other migration in the same app).
+
+When :djadmin:`migrate` is run with the ``--fake-initial`` argument, these initial
+migrations are treated specially. For an initial migration that creates a one
```
Comment (MarkusH) [inline docs/topics/migrations.txt]: "... creates one" -- there is an "a" too much.

File: docs/topics/migrations.txt [hunk 14]
```diff
@@ -276,6 +276,36 @@ class to make it importable::
...
+or more tables (``CreateModel`` operation), Django will check if those tables
```
Comment (MarkusH) [inline docs/topics/migrations.txt]: "check *that all* those tables"

File: docs/topics/migrations.txt [hunk 15]
```diff
@@ -276,6 +276,36 @@ class to make it importable::
...
+already exist in the database, and fake-apply the migration if so. Similarly,
+for an initial migration that adds one or more fields (``AddField`` operation),
+Django will check if those columns already exist in the database, and
```
Comment (MarkusH) [inline docs/topics/migrations.txt]: "check *that all their respective* columns"

File: docs/topics/migrations.txt [hunk 16]
```diff
@@ -276,6 +276,36 @@ class to make it importable::
...
+fake-apply the migration if so. When :djadminopt:`--fake-initial` is not used,
```
Comment (MarkusH) [inline docs/topics/migrations.txt]: Maybe rephrase to "Without --fake-initial, ..."

File: django/db/migrations/executor.py [hunk 17]
```diff
@@ -166,16 +166,21 @@ def detect_soft_applied(self, project_state, migration):
         tables it would create exist. This is intended only for use
```
Comment (MarkusH) [inline django/db/migrations/executor.py]: "tables or columns it would create"

File: django/db/migrations/executor.py [hunk 18]
```diff
@@ -166,16 +166,21 @@ def detect_soft_applied(self, project_state, migration):
...
         on initial migrations (as it only looks for CreateModel).
```
Comment (MarkusH) [inline django/db/migrations/executor.py]: "CreateModel and AddField"

File: django/db/migrations/executor.py [hunk 19]
```diff
@@ -186,6 +191,20 @@ def detect_soft_applied(self, project_state, migration):
                 if model._meta.db_table not in self.connection.introspection.table_names(self.connection.cursor()):
                     return False, project_state
                 found_create_migration = True
-        # If we get this far and we found at least one CreateModel migration,
+
+            elif isinstance(operation, migrations.AddField):
+                model = apps.get_model(migration.app_label, operation.model_name)
+                if model._meta.swapped:
```
Comment (MarkusH) [inline django/db/migrations/executor.py]: I think we need to ignore models with `managed=False` or `proxy=True` here as they never receive any database level operations anyways.

File: django/db/migrations/autodetector.py [hunk 20]
```diff
@@ -33,6 +33,7 @@ def __init__(self, from_state, to_state, questioner=None):
         self.from_state = from_state
         self.to_state = to_state
         self.questioner = questioner or MigrationQuestioner()
+        self.existing_apps = set([app for app, model in from_state.models])
```
Comment (timgraham) [inline django/db/migrations/autodetector.py]: use set comprehension: `{app for app, ... }`

File: django/db/migrations/executor.py [hunk 21]
```diff
@@ -163,19 +163,24 @@ def unapply_migration(self, state, migration, fake=False):
...
-        on initial migrations (as it only looks for CreateModel).
+        tables or columns it would create exist. This is intended only for use
+        on initial migrations (as it only looks for CreateModel and AddField).
         """
-        # Bail if the migration isn't the first one in its app
-        if [name for app, name in migration.dependencies if app == migration.app_label]:
+
+        # Bail if it's NOT an initial migration
+        if migration.initial is None:
+            if [name for app, name in migration.dependencies if app == migration.app_label]:
```
Comment (timgraham) [inline django/db/migrations/executor.py]: use: `any(name for app, name ... )`

File: django/db/migrations/migration.py [hunk 22]
```diff
@@ -39,6 +39,12 @@ class Migration(object):
     # are not applied.
     replaces = []
 
+    # Initial migration, should be silently skipped on --fake-initial if table or fields already
```
Comment (timgraham) [inline django/db/migrations/migration.py]: please wrap comments at 79 chars

**Comment by akulakov** [comment]
Thanks Tim, applied these changes..

File: django/db/migrations/migration.py [hunk 23]
```diff
@@ -41,6 +41,17 @@ class Migration(object):
...
 
+    # Initial migration, should be silently skipped on --fake-initial if table
+    # or fields already exists; if None, fallback to the old way of checking if
+    # there are dependencies to tell if db introspection needs to be done; if
+    # True, always perform introspection; if False, never perform
+    # introspection.
+    initial = None
+
+    # Error class which is raised when a migration is irreversible
+    class IrreversibleError(RuntimeError):
```
Comment (timgraham) [inline django/db/migrations/migration.py]: rebase error added this?

File: django/db/migrations/executor.py [hunk 24]
```diff
@@ -199,9 +204,27 @@ def detect_soft_applied(self, project_state, migration):
...
+                    # We have to fetch the model to test with from the
+                    # main app cache, as it's not a direct dependency.
+                    model = global_apps.get_model(model._meta.swapped)
+                if model._meta.proxy or not model._meta.managed:
+                    continue
+
+                table = model._meta.db_table
+                db_field = model._meta.get_field(operation.name).column
+                fields = self.connection.introspection.get_table_description(self.connection.cursor(), table)
+                if db_field not in [f.name for f in fields]:
```
Comment (timgraham) [inline django/db/migrations/executor.py]: `all(f.name != db_field for f in fields)` (not sure if it's bettter/worse than the current version)

File: django/db/migrations/executor.py [hunk 25]
```diff
@@ -179,19 +179,24 @@ def unapply_migration(self, state, migration, fake=False):
...
         """
         Tests whether a migration has been implicitly applied - that the
-        tables it would create exist. This is intended only for use
-        on initial migrations (as it only looks for CreateModel).
+        tables or columns it would create exist. This is intended only for use
+        on initial migrations (as it only looks for CreateModel and AddField).
         """
-        # Bail if the migration isn't the first one in its app
-        if [name for app, name in migration.dependencies if app == migration.app_label]:
+
```
Comment (timgraham) [inline django/db/migrations/executor.py]: no blank line after docstring

File: docs/topics/migrations.txt [hunk 26]
```diff
@@ -276,6 +276,37 @@ class to make it importable::
 Please refer to the notes about :ref:`historical-models` in migrations to see
 the implications that come along.
 
+.. _initial-property:
```
Comment (timgraham) [inline docs/topics/migrations.txt]: migrations-initial-property
+ Please add a link from the release notes.

File: docs/topics/migrations.txt [hunk 27]
```diff
@@ -276,6 +276,37 @@ class to make it importable::
...
+or more.
+
+Initial migrations are marked with an ``initial = True`` class attribute on the
+migration class. If no initial class attribute is found, a migration will be
+considered "initial" if it is the first migration in the app (i.e. it has no
+dependencies on any other migration in the same app).
+
+When :djadmin:`migrate` is run with the ``--fake-initial`` argument, these initial
+migrations are treated specially. For an initial migration that creates one
+or more tables (``CreateModel`` operation), Django will check that all those tables
```
Comment (timgraham) [inline docs/topics/migrations.txt]: all of those

File: docs/topics/migrations.txt [hunk 28]
```diff
@@ -276,6 +276,37 @@ class to make it importable::
...
+already exist in the database, and fake-apply the migration if so. Similarly,
```
Comment (timgraham) [inline docs/topics/migrations.txt]: no comma needed before and

File: docs/topics/migrations.txt [hunk 29]
```diff
@@ -276,6 +276,37 @@ class to make it importable::
...
+for an initial migration that adds one or more fields (``AddField`` operation),
+Django will check that all their respective columns already exist in the
```
Comment (timgraham) [inline docs/topics/migrations.txt]: all of their

File: docs/topics/migrations.txt [hunk 30]
```diff
@@ -276,6 +276,37 @@ class to make it importable::
...
+database, and fake-apply the migration if so. Without
```
Comment (timgraham) [inline docs/topics/migrations.txt]: chop comma as above

File: django/db/migrations/migration.py [hunk 31]
```diff
@@ -41,6 +41,17 @@ class Migration(object):
     # are not applied.
     replaces = []
 
+    # Initial migration, should be silently skipped on --fake-initial if table
+    # or fields already exists; if None, fallback to the old way of checking if
```
Comment (timgraham) [inline django/db/migrations/migration.py]: if None, check if the migration has any dependencies to determine if ...

File: django/db/migrations/migration.py [hunk 32]
```diff
@@ -41,6 +41,17 @@ class Migration(object):
     # are not applied.
     replaces = []
 
+    # Initial migration, should be silently skipped on --fake-initial if table
```
Comment (timgraham) [inline django/db/migrations/migration.py]: Is this an initial migration? Initial migrations are skipped on --fake-initial...

File: django/db/migrations/writer.py [hunk 33]
```diff
@@ -509,7 +513,8 @@ def serialize(cls, value):
 %(imports)s
 
 class Migration(migrations.Migration):
-%(replaces_str)s
+%(replaces_str)s%(initial_str)s
+
```
Comment (timgraham) [inline django/db/migrations/writer.py]: extra newline if initial_str is empty -- I think iniitial_str should have a \n at the end like replaces_str does

**Comment by timgraham** [comment]
Should we have a test with `initial=False`. Did you see Markus's comment on the ticket about issues with squashed migrations?

File: django/db/migrations/executor.py [hunk 34]
```diff
@@ -199,9 +204,27 @@ def detect_soft_applied(self, project_state, migration):
                     # We have to fetch the model to test with from the
                     # main app cache, as it's not a direct dependency.
                     model = global_apps.get_model(model._meta.swapped)
+                if model._meta.proxy or not model._meta.managed:
+                    continue
                 if model._meta.db_table not in self.connection.introspection.table_names(self.connection.cursor()):
                     return False, project_state
                 found_create_migration = True
```
Comment (MarkusH) [inline django/db/migrations/executor.py]: I'd rename that to `found_create_model_migration`

File: django/db/migrations/executor.py [hunk 35]
```diff
@@ -199,9 +204,27 @@ def detect_soft_applied(self, project_state, migration):
...
+                    model = global_apps.get_model(model._meta.swapped)
+                if model._meta.proxy or not model._meta.managed:
+                    continue
+
+                table = model._meta.db_table
+                db_field = model._meta.get_field(operation.name).column
+                fields = self.connection.introspection.get_table_description(self.connection.cursor(), table)
+                if db_field not in [f.name for f in fields]:
+                    return False, project_state
+                found_create_migration = True
```
Comment (MarkusH) [inline django/db/migrations/executor.py]: I'd change that to `found_add_field_migration`

File: django/db/migrations/executor.py [hunk 36]
```diff
@@ -199,9 +204,27 @@ def detect_soft_applied(self, project_state, migration):
...
+        # If we get this far and we found at least one CreateModel or AddField migration,
         # the migration is considered implicitly applied.
         return found_create_migration, after_state
```
Comment (MarkusH) [inline django/db/migrations/executor.py]: ... and change this to `return (found_create_model_migration or found_add_field_migration), after_state`

File: django/db/migrations/migration.py (hunk 23 above)
Comment (MarkusH) [inline django/db/migrations/migration.py]: yes

File: django/db/migrations/executor.py [hunk 37]
```diff
@@ -179,19 +179,24 @@ def unapply_migration(self, state, migration, fake=False):
...
             return False, project_state
+
         if project_state is None:
             after_state = self.loader.project_state((migration.app_label, migration.name), at_end=True)
         else:
             after_state = migration.mutate_state(project_state)
         apps = after_state.apps
         found_create_migration = False
-        # Make sure all create model are done
+        # Make sure all create model and add field operations are done
```
Comment (MarkusH) [inline django/db/migrations/executor.py]: Could you replace "create model" and "add field" with "CreateModel" and "AddField" respectively, please.

File: django/db/migrations/executor.py (hunk 19 above)
Comment (MarkusH) [inline django/db/migrations/executor.py]: Thanks for addressing this.

File: django/db/migrations/executor.py (hunk 24 above)
Comment (MarkusH) [inline django/db/migrations/executor.py]: I think the current version is clear. You could change the `[...]` to `(...)` though

Comment (akulakov) [inline django/db/migrations/executor.py]: I think either current version or `not any(f.name == db_field for f in fields)` are clearer than `all()`.

File: docs/topics/migrations.txt [hunk 38]
```diff
@@ -276,6 +276,37 @@ class to make it importable::
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. versionadded:: 1.9
+
+The "initial migrations" for an app are the migrations that create the first
+version of that app's tables. Usually an app will have just one initial
+migration, but in some cases of complex model interdependencies it may have two
+or more.
+
+Initial migrations are marked with an ``initial = True`` class attribute on the
```
Comment (akulakov) [inline docs/topics/migrations.txt]: Perhaps 'class property' here is better because it's more specific?

Comment (timgraham) [inline docs/topics/migrations.txt]: I believe [class attribute](http://www.diveintopython.net/object_oriented_framework/class_attributes.html) is the correct terminology. Properties are different from attributes.

Comment (akulakov) [inline docs/topics/migrations.txt]: My mistake, I misremembered the exact meaning of property.

**Comment by akulakov** [comment]
I will work on the squashed migrations issue this week.

**Comment by akulakov** [comment]
Hi I've added initial=True setting to squashed migrations, along with a unit test. Seems to work
well -- I've read through squashing docs but I haven't worked in that area before (or even used squashing before..).

**Comment by akulakov** [comment]
There's a bunch of examples in the main migrations doc that show a migration file with a '0001_initial' name, I think I should update all of them to show the attribute as well?

**Comment by carljm** [comment]
Yes, squashed migrations generated by Django should get `initial = True`. It's not inherently required for a migration with `replaces` to be initial, but the `squashmigrations` management command is only capable of creating initial ones, and that's where `initial = True` is being set.

And yes, any initial migration files whose full contents we show in the docs should now be shown with `initial=True`; good catch!

**Comment by akulakov** [comment]
@carljm thanks! I've added the changes to the migrations doc.

**Comment by akulakov** [comment]
@timgraham not sure if we should test initial=False -- we would be essentially testing only these two lines in migrations/executor.py, lines 200-201:

        elif migration.initial is False:
             return False, project_state

**Comment by MarkusH** [comment]
Yes, I want these to lines to be tested, too.

**Comment by akulakov** [comment]
I've added the test for `initial = False`, sorry for taking a while..

File: django/db/migrations/executor.py [hunk 39]
```diff
@@ -190,19 +190,24 @@ def check_replacements(self):
...
         """
         Tests whether a migration has been implicitly applied - that the
-        tables it would create exist. This is intended only for use
-        on initial migrations (as it only looks for CreateModel).
+        tables or columns it would create exist. This is intended only for use
+        on initial migrations (as it only looks for CreateModel and AddField).
         """
-        # Bail if the migration isn't the first one in its app
-        if [name for app, name in migration.dependencies if app == migration.app_label]:
+        # Bail if it's NOT an initial migration
```
Comment (MarkusH) [inline django/db/migrations/executor.py]: Can you move this comment into the `elif` branch and the one you removed into the `if any()`, please.

File: docs/topics/migrations.txt (hunk 26 above)
Comment (MarkusH) [inline docs/topics/migrations.txt]: This needs an update

Comment (akulakov) [inline docs/topics/migrations.txt]: Sorry, what needs to be updated here?

Comment (MarkusH) [inline docs/topics/migrations.txt]: Please change `initial-property` to `migration-initial-property` (slugify the section title)

Comment (akulakov) [inline docs/topics/migrations.txt]: Ah, makes sense! Pushed the update..

**Comment by MarkusH** [comment]
docs:
```
docs/topics/migrations.txt:799: WARNING: undefined label: initial-property (if the link has no caption the label must precede a section header)
```

flake8: error on master

default:
```python
======================================================================
ERROR [0.000s]: test_migrate_initial_false (migrations.test_commands.MigrateTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jenkins/workspace/pull-requests-trusty/database/sqlite3/label/trusty-pr/python/python2.7/django/test/utils.py", line 182, in inner
    return test_func(*args, **kwargs)
  File "/home/jenkins/workspace/pull-requests-trusty/database/sqlite3/label/trusty-pr/python/python2.7/tests/migrations/test_commands.py", line 62, in test_migrate_initial_false
    call_command("migrate", "migrations", "0001", verbosity=0)
  File "/home/jenkins/workspace/pull-requests-trusty/database/sqlite3/label/trusty-pr/python/python2.7/django/core/management/__init__.py", line 118, in call_command
    return command.execute(*args, **defaults)
  File "/home/jenkins/workspace/pull-requests-trusty/database/sqlite3/label/trusty-pr/python/python2.7/django/core/management/base.py", line 401, in execute
    output = self.handle(*args, **options)
  File "/home/jenkins/workspace/pull-requests-trusty/database/sqlite3/label/trusty-pr/python/python2.7/django/core/management/commands/migrate.py", line 107, in handle
    "App '%s' does not have migrations." % app_label
CommandError: App 'migrations' does not have migrations.

----------------------------------------------------------------------
```

**Comment by akulakov** [comment]
Fixed the ref link.
Not sure why CI was picking up flake8 error which wasn't in my changesets.
The test error was caused by the older update, I should have added a comment after I fixed it..

**Comment by timgraham** [comment]
Looks good, I have only some minor cosmetic edits: http://dpaste.com/211XKE4
I see that Markus suggested it, but I am not sure this needs to be two commits. In any case, it's a bit confusing to have two commits that both say "Fixed #....". I guess the second one could be modified to say "Refs #..." and some other description, but I'll leave it up to you, Andrei.

**Comment by akulakov** [comment]
Sorry, had too much work this week to look at this sooner.. @timgraham your updates look good, can you go ahead and apply them, or I should do that?

I've changed the title and description of second commit as you recommend, please see if it looks good.

Thanks for the reviews & updates!

**Comment by timgraham** [comment]
merged in db97a8849519a3933bf4abd2184efd68ebc21965, thanks.
